xmodule/docs/decisions/0004-evaluate-wasm-safe-exec.rst

```diff
@@ -0,0 +1,107 @@
+
+Flexibility of Additional Runtimes
+----------------------------------
+
+When the sandboxed runtime environment is just a Wasm file, it makes it much easier for us to support multiple runtimes with the same service and add new ones as necessary. These runtimes can even be in different languages.
+
+Operations
+----------
+
+The Wasmtime library gives a more precise accounting of the resources being used by allotting computation budget "fuel" and measuring how much a given Wasm program has used. This is more precise than the limits that AppArmor uses, which can sometimes result in problems that work fine most of the time but start failing under elevated system load. The fact that we have precise measurements of how much fuel is being used also means that we can surface this information to content authors, so that they can be aware of the problem.
+
+Cross-Platform Support
+----------------------
+
+AppArmor only works in Linux, and cannot run on Linux guest Docker containers on a macOS host machine. This makes it difficult for developers running macOS to develop and test features related to sandboxed code execution. WebAssembly can run on any major operating system.
+
+Decisions
+*********
+
+#. We will create a new library in a new repo (``wasm-safe-exec``) for safely executing instructor-authored code using WebAssembly as the execution mechanism.
+
+   * The library will use ``wasmtime-py`` for Wasm code execution.
+   * The library must be runnable in edx-platform, meaning that it still needs to support Python 3.8 until edx-platform has been upgraded to 3.11.
+   * The library must be able to simultaneously support multiple code execution runtime environments.
+   * The library must not make ``edx-platform`` specific assumptions–e.g. ``UsageKeys``. It should be generally reusable for untrusted code execution by other projects.
+
+#. We will use the WebAssembly Component Model and create a Wasm Interface Type (WIT) file defining the interface between the host library and guest code execution runtime environments.
+
+   * It should be possible for third parties to easily create their own runtime environments.
+
+#. We will create two simple runtimes to start: Python 3.11 and JavaScript ES14.
+
+   * Runtime binaries will be created with `componentize-py <https://component-model.bytecodealliance.org/language-support/python.html>`_ for Python and and `jco <https://component-model.bytecodealliance.org/language-support/javascript.html>`_ for JavaScript.
+   * Python 3.11 is the first version of Python that officially supports compiling to ``wasm32-wasi``, which is why it's the one we'll start with.
+   * The Python runtime must be able to accept a course's ``python_lib.zip`` file.
```

Comment:
Great, Would this work the same way as before with codejail? e.g. adding the `python_lib.zip` through course assets?

Reply:
Yes. I do have other ways I want to enhance the code authoring experience in the next year, like:

* Allow for multiple `python_lib.zip`-type libraries that are separately versioned and sharable via libraries.
* Let people make a separate Python file instead of having to embed it in the XML with a `src` attribute in the script tag.
* A feedback meter that shows how much CPU and memory are used when executing the problem.

But that's something that would come after a couple of enhancements to v2 content libraries. For this purposes of this, I want to shoot for maximum backwards compatibility with codejail.